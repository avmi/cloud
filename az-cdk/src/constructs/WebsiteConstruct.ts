import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import {
  PublicHostedZone,
  ARecord,
  AaaaRecord,
  MxRecord,
  RecordTarget,
  IPublicHostedZone,
} from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import { Bucket } from '@aws-cdk/aws-s3';
import { CloudFrontWebDistribution, PriceClass, AliasConfiguration } from '@aws-cdk/aws-cloudfront';
import { VerifyDomainConstruct } from '../custom-resources/VerifyDomainConstruct';

export interface IWebsiteProps {
  domain: string; // e.g. mydomain.com
  prefix?: string; // e.g. 'app' for deploying app.mydomain.com ('www' will not be created then)
  skipMX?: boolean; // skip setting MX record set
  skipVerification?: boolean; // skip domain verificaton
  comment?: string; // e.g. My awesome domain
  hostedZoneId?: string; // use an existend hosted zone, otherwise create a new (public) one
  certificateArn?: string; // Arn of an existent and VALID certificate. Use empty or 'null' to skip use of certificate
  errorCodes?: number[]; // default = 403, 404
  errorRoute?: string; // default = /index.html
}

// Website creates bucket, cloudformation, certificate, and route53 data for a website
export class WebsiteConstruct extends Construct {
  readonly cloudfrontDistributionId: string;

  constructor(scope: Construct, id: string, props: IWebsiteProps) {
    super(scope, id);

    const certificateArn = props.certificateArn || 'null';
    const prefixedDomain = `${props.prefix ? props.prefix + '.' : ''}${props.domain}`;
    const errorCodes = props.errorCodes || [403, 404];
    const errorRoute = props.errorRoute || '/index.html';

    let zone: IPublicHostedZone;
    if (props.hostedZoneId) {
      zone = PublicHostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
        hostedZoneId: props.hostedZoneId,
        zoneName: props.domain,
      });
    } else {
      zone = new PublicHostedZone(this, 'HostedZone', {
        zoneName: props.domain,
        comment: props.comment,
        caaAmazon: true,
      });
    }

    const bucket = new Bucket(this, 'Bucket', {
      bucketName: `${prefixedDomain}-${id.toLowerCase()}`,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    let aliasConfiguration: AliasConfiguration | undefined;
    if (certificateArn !== 'null') {
      aliasConfiguration = {
        acmCertRef: certificateArn,
        names: [prefixedDomain],
      };
      if (!props.prefix) {
        aliasConfiguration.names.push(`www.${props.domain}`);
      }
    }

    const distribution = new CloudFrontWebDistribution(this, 'Distribution', {
      aliasConfiguration,
      defaultRootObject: undefined,
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: bucket,
          },
          behaviors: [
            {
              isDefaultBehavior: true,
            },
          ],
        },
      ],
      priceClass: PriceClass.PRICE_CLASS_ALL,
      errorConfigurations: errorCodes.map(code => ({
        errorCachingMinTtl: 300,
        errorCode: code,
        responseCode: 200,
        responsePagePath: errorRoute,
      })),
    });

    this.cloudfrontDistributionId = distribution.distributionId;

    const target = new CloudFrontTarget(distribution);

    // prefixed
    if (props.prefix) {
      new ARecord(this, 'A', {
        zone,
        recordName: props.prefix,
        target: RecordTarget.fromAlias(target),
      });

      new AaaaRecord(this, 'AAA', {
        zone,
        recordName: props.prefix,
        target: RecordTarget.fromAlias(target),
      });

      // standard
    } else {
      new ARecord(this, 'A', {
        zone,
        target: RecordTarget.fromAlias(target),
      });

      new AaaaRecord(this, 'AAAA', {
        zone,
        target: RecordTarget.fromAlias(target),
      });

      new ARecord(this, 'WWW', {
        zone,
        recordName: 'www',
        target: RecordTarget.fromAlias(target),
      });

      if (!props.skipMX) {
        new MxRecord(this, 'MX', {
          zone,
          values: [
            {
              priority: 10,
              hostName: `inbound-smtp.${Aws.REGION}.amazonaws.com`,
            },
          ],
        });
      }

      if (!props.skipVerification) {
        new VerifyDomainConstruct(this, 'VerifyDomain', {
          hostedZoneId: zone.hostedZoneId,
          domain: props.domain,
          certificateArn,
        });
      }
    }
  }
}
