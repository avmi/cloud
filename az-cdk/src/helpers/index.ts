export * from './apiGatewayHostedZoneId';
export * from './envars2cdk';
export * from './npmCommands';
export * from './ssmSecret';
export * from './synthApp';
