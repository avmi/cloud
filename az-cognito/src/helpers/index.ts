export * from './flattenAttributes';
export * from './email2username';
