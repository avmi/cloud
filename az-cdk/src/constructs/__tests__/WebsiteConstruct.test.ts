import { App, Stack, CfnOutput } from '@aws-cdk/core';
import { WebsiteConstruct } from '../WebsiteConstruct';
import { synthAppString } from '../../helpers/synthApp';

describe('WebsiteConstruct', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new WebsiteConstruct(stack, 'Website', {
      domain: 'mydomain.com',
      comment: 'My nice website',
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });

  it('synthetizes properly using certificateArn', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    const website = new WebsiteConstruct(stack, 'Website', {
      domain: 'mydomain.com',
      comment: 'My nice website',
      certificateArn: 'arn:aws:acm:REGION:ACCOUNT:certificate/12345-678-9012345',
    });
    new CfnOutput(stack, 'CloudFrontDistributionId', {
      value: website.cloudfrontDistributionId,
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});

describe('WebsiteConstruct (prefixed)', () => {
  it('synthetizes properly using certificateArn', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new WebsiteConstruct(stack, 'Website', {
      domain: 'mydomain.com',
      prefix: 'app',
      comment: 'My app website',
      certificateArn: 'arn:aws:acm:REGION:ACCOUNT:certificate/12345-678-9012345',
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});
