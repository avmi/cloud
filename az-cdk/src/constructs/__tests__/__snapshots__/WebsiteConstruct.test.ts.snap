// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebsiteConstruct synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"WebsiteHostedZoneD4FAEADF\\": {
      \\"Type\\": \\"AWS::Route53::HostedZone\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"HostedZoneConfig\\": {
          \\"Comment\\": \\"My nice website\\"
        }
      }
    },
    \\"WebsiteHostedZoneCaaAmazonF67E60BB\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"CAA\\",
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        },
        \\"ResourceRecords\\": [
          \\"0 issue \\\\\\"amazon.com\\\\\\"\\"
        ],
        \\"TTL\\": \\"1800\\"
      }
    },
    \\"WebsiteBucket4326D7C2\\": {
      \\"Type\\": \\"AWS::S3::Bucket\\",
      \\"Properties\\": {
        \\"BucketName\\": \\"mydomain.com-website\\"
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    },
    \\"WebsiteDistributionCFDistributionD4D90A52\\": {
      \\"Type\\": \\"AWS::CloudFront::Distribution\\",
      \\"Properties\\": {
        \\"DistributionConfig\\": {
          \\"CustomErrorResponses\\": [
            {
              \\"ErrorCachingMinTTL\\": 300,
              \\"ErrorCode\\": 403,
              \\"ResponseCode\\": 200,
              \\"ResponsePagePath\\": \\"/index.html\\"
            },
            {
              \\"ErrorCachingMinTTL\\": 300,
              \\"ErrorCode\\": 404,
              \\"ResponseCode\\": 200,
              \\"ResponsePagePath\\": \\"/index.html\\"
            }
          ],
          \\"DefaultCacheBehavior\\": {
            \\"AllowedMethods\\": [
              \\"GET\\",
              \\"HEAD\\"
            ],
            \\"CachedMethods\\": [
              \\"GET\\",
              \\"HEAD\\"
            ],
            \\"Compress\\": true,
            \\"ForwardedValues\\": {
              \\"Cookies\\": {
                \\"Forward\\": \\"none\\"
              },
              \\"QueryString\\": false
            },
            \\"TargetOriginId\\": \\"origin1\\",
            \\"ViewerProtocolPolicy\\": \\"redirect-to-https\\"
          },
          \\"DefaultRootObject\\": \\"index.html\\",
          \\"Enabled\\": true,
          \\"HttpVersion\\": \\"http2\\",
          \\"IPV6Enabled\\": true,
          \\"Origins\\": [
            {
              \\"DomainName\\": {
                \\"Fn::GetAtt\\": [
                  \\"WebsiteBucket4326D7C2\\",
                  \\"RegionalDomainName\\"
                ]
              },
              \\"Id\\": \\"origin1\\",
              \\"S3OriginConfig\\": {}
            }
          ],
          \\"PriceClass\\": \\"PriceClass_All\\",
          \\"ViewerCertificate\\": {
            \\"CloudFrontDefaultCertificate\\": true
          }
        }
      }
    },
    \\"WebsiteA6C1FD492\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"A\\",
        \\"AliasTarget\\": {
          \\"DNSName\\": {
            \\"Fn::GetAtt\\": [
              \\"WebsiteDistributionCFDistributionD4D90A52\\",
              \\"DomainName\\"
            ]
          },
          \\"HostedZoneId\\": \\"Z2FDTNDATAQYW2\\"
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        }
      }
    },
    \\"WebsiteAAAAA3AA9709\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"AAAA\\",
        \\"AliasTarget\\": {
          \\"DNSName\\": {
            \\"Fn::GetAtt\\": [
              \\"WebsiteDistributionCFDistributionD4D90A52\\",
              \\"DomainName\\"
            ]
          },
          \\"HostedZoneId\\": \\"Z2FDTNDATAQYW2\\"
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        }
      }
    },
    \\"WebsiteWWWE5086576\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"www.mydomain.com.\\",
        \\"Type\\": \\"A\\",
        \\"AliasTarget\\": {
          \\"DNSName\\": {
            \\"Fn::GetAtt\\": [
              \\"WebsiteDistributionCFDistributionD4D90A52\\",
              \\"DomainName\\"
            ]
          },
          \\"HostedZoneId\\": \\"Z2FDTNDATAQYW2\\"
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        }
      }
    },
    \\"WebsiteMX1A2D5733\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"MX\\",
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        },
        \\"ResourceRecords\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"10 inbound-smtp.\\",
                {
                  \\"Ref\\": \\"AWS::Region\\"
                },
                \\".amazonaws.com\\"
              ]
            ]
          }
        ],
        \\"TTL\\": \\"1800\\"
      }
    },
    \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\": {
      \\"Type\\": \\"AWS::IAM::Role\\",
      \\"Properties\\": {
        \\"AssumeRolePolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sts:AssumeRole\\",
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"lambda.amazonaws.com\\"
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"ManagedPolicyArns\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:\\",
                {
                  \\"Ref\\": \\"AWS::Partition\\"
                },
                \\":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\"
              ]
            ]
          }
        ]
      }
    },
    \\"WebsiteVerifyDomainFunctionServiceRoleDefaultPolicy90A097B8\\": {
      \\"Type\\": \\"AWS::IAM::Policy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"acm:DescribeCertificate\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:VerifyDomainIdentity\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:VerifyDomainDkim\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:GetIdentityVerificationAttributes\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:GetIdentityDkimAttributes\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:DeleteIdentity\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"route53:ChangeResourceRecordSets\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": {
                \\"Fn::Join\\": [
                  \\"\\",
                  [
                    \\"arn:aws:route53:::hostedzone/\\",
                    {
                      \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
                    }
                  ]
                ]
              }
            },
            {
              \\"Action\\": \\"route53:GetChange\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": {
                \\"Fn::Join\\": [
                  \\"\\",
                  [
                    \\"arn:aws:route53:::hostedzone/\\",
                    {
                      \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
                    }
                  ]
                ]
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"PolicyName\\": \\"WebsiteVerifyDomainFunctionServiceRoleDefaultPolicy90A097B8\\",
        \\"Roles\\": [
          {
            \\"Ref\\": \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\"
          }
        ]
      }
    },
    \\"WebsiteVerifyDomainFunction9978DE02\\": {
      \\"Type\\": \\"AWS::Lambda::Function\\",
      \\"Properties\\": {
        \\"Code\\": {
          \\"S3Bucket\\": {
            \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\"
          },
          \\"S3Key\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                },
                {
                  \\"Fn::Select\\": [
                    1,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        \\"Handler\\": \\"index.verifyDomain\\",
        \\"Role\\": {
          \\"Fn::GetAtt\\": [
            \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\",
            \\"Arn\\"
          ]
        },
        \\"Runtime\\": \\"nodejs8.10\\",
        \\"Timeout\\": 60
      },
      \\"DependsOn\\": [
        \\"WebsiteVerifyDomainFunctionServiceRoleDefaultPolicy90A097B8\\",
        \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\"
      ]
    },
    \\"WebsiteVerifyDomainB2C2439A\\": {
      \\"Type\\": \\"AWS::CloudFormation::CustomResource\\",
      \\"Properties\\": {
        \\"ServiceToken\\": {
          \\"Fn::GetAtt\\": [
            \\"WebsiteVerifyDomainFunction9978DE02\\",
            \\"Arn\\"
          ]
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        },
        \\"DomainName\\": \\"mydomain.com\\",
        \\"CertificateArn\\": \\"null\\"
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    }
  },
  \\"Parameters\\": {
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 bucket for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 key for asset version \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168ArtifactHashDB14117A\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"Artifact hash for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    }
  }
}"
`;

exports[`WebsiteConstruct synthetizes properly using certificateArn 1`] = `
"{
  \\"Resources\\": {
    \\"WebsiteHostedZoneD4FAEADF\\": {
      \\"Type\\": \\"AWS::Route53::HostedZone\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"HostedZoneConfig\\": {
          \\"Comment\\": \\"My nice website\\"
        }
      }
    },
    \\"WebsiteHostedZoneCaaAmazonF67E60BB\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"CAA\\",
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        },
        \\"ResourceRecords\\": [
          \\"0 issue \\\\\\"amazon.com\\\\\\"\\"
        ],
        \\"TTL\\": \\"1800\\"
      }
    },
    \\"WebsiteBucket4326D7C2\\": {
      \\"Type\\": \\"AWS::S3::Bucket\\",
      \\"Properties\\": {
        \\"BucketName\\": \\"mydomain.com-website\\"
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    },
    \\"WebsiteDistributionCFDistributionD4D90A52\\": {
      \\"Type\\": \\"AWS::CloudFront::Distribution\\",
      \\"Properties\\": {
        \\"DistributionConfig\\": {
          \\"Aliases\\": [
            \\"mydomain.com\\",
            \\"www.mydomain.com\\"
          ],
          \\"CustomErrorResponses\\": [
            {
              \\"ErrorCachingMinTTL\\": 300,
              \\"ErrorCode\\": 403,
              \\"ResponseCode\\": 200,
              \\"ResponsePagePath\\": \\"/index.html\\"
            },
            {
              \\"ErrorCachingMinTTL\\": 300,
              \\"ErrorCode\\": 404,
              \\"ResponseCode\\": 200,
              \\"ResponsePagePath\\": \\"/index.html\\"
            }
          ],
          \\"DefaultCacheBehavior\\": {
            \\"AllowedMethods\\": [
              \\"GET\\",
              \\"HEAD\\"
            ],
            \\"CachedMethods\\": [
              \\"GET\\",
              \\"HEAD\\"
            ],
            \\"Compress\\": true,
            \\"ForwardedValues\\": {
              \\"Cookies\\": {
                \\"Forward\\": \\"none\\"
              },
              \\"QueryString\\": false
            },
            \\"TargetOriginId\\": \\"origin1\\",
            \\"ViewerProtocolPolicy\\": \\"redirect-to-https\\"
          },
          \\"DefaultRootObject\\": \\"index.html\\",
          \\"Enabled\\": true,
          \\"HttpVersion\\": \\"http2\\",
          \\"IPV6Enabled\\": true,
          \\"Origins\\": [
            {
              \\"DomainName\\": {
                \\"Fn::GetAtt\\": [
                  \\"WebsiteBucket4326D7C2\\",
                  \\"RegionalDomainName\\"
                ]
              },
              \\"Id\\": \\"origin1\\",
              \\"S3OriginConfig\\": {}
            }
          ],
          \\"PriceClass\\": \\"PriceClass_All\\",
          \\"ViewerCertificate\\": {
            \\"AcmCertificateArn\\": \\"arn:aws:acm:REGION:ACCOUNT:certificate/12345-678-9012345\\",
            \\"SslSupportMethod\\": \\"sni-only\\"
          }
        }
      }
    },
    \\"WebsiteA6C1FD492\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"A\\",
        \\"AliasTarget\\": {
          \\"DNSName\\": {
            \\"Fn::GetAtt\\": [
              \\"WebsiteDistributionCFDistributionD4D90A52\\",
              \\"DomainName\\"
            ]
          },
          \\"HostedZoneId\\": \\"Z2FDTNDATAQYW2\\"
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        }
      }
    },
    \\"WebsiteAAAAA3AA9709\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"AAAA\\",
        \\"AliasTarget\\": {
          \\"DNSName\\": {
            \\"Fn::GetAtt\\": [
              \\"WebsiteDistributionCFDistributionD4D90A52\\",
              \\"DomainName\\"
            ]
          },
          \\"HostedZoneId\\": \\"Z2FDTNDATAQYW2\\"
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        }
      }
    },
    \\"WebsiteWWWE5086576\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"www.mydomain.com.\\",
        \\"Type\\": \\"A\\",
        \\"AliasTarget\\": {
          \\"DNSName\\": {
            \\"Fn::GetAtt\\": [
              \\"WebsiteDistributionCFDistributionD4D90A52\\",
              \\"DomainName\\"
            ]
          },
          \\"HostedZoneId\\": \\"Z2FDTNDATAQYW2\\"
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        }
      }
    },
    \\"WebsiteMX1A2D5733\\": {
      \\"Type\\": \\"AWS::Route53::RecordSet\\",
      \\"Properties\\": {
        \\"Name\\": \\"mydomain.com.\\",
        \\"Type\\": \\"MX\\",
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        },
        \\"ResourceRecords\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"10 inbound-smtp.\\",
                {
                  \\"Ref\\": \\"AWS::Region\\"
                },
                \\".amazonaws.com\\"
              ]
            ]
          }
        ],
        \\"TTL\\": \\"1800\\"
      }
    },
    \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\": {
      \\"Type\\": \\"AWS::IAM::Role\\",
      \\"Properties\\": {
        \\"AssumeRolePolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sts:AssumeRole\\",
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"lambda.amazonaws.com\\"
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"ManagedPolicyArns\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:\\",
                {
                  \\"Ref\\": \\"AWS::Partition\\"
                },
                \\":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\"
              ]
            ]
          }
        ]
      }
    },
    \\"WebsiteVerifyDomainFunctionServiceRoleDefaultPolicy90A097B8\\": {
      \\"Type\\": \\"AWS::IAM::Policy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"acm:DescribeCertificate\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:VerifyDomainIdentity\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:VerifyDomainDkim\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:GetIdentityVerificationAttributes\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:GetIdentityDkimAttributes\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:DeleteIdentity\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"route53:ChangeResourceRecordSets\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": {
                \\"Fn::Join\\": [
                  \\"\\",
                  [
                    \\"arn:aws:route53:::hostedzone/\\",
                    {
                      \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
                    }
                  ]
                ]
              }
            },
            {
              \\"Action\\": \\"route53:GetChange\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": {
                \\"Fn::Join\\": [
                  \\"\\",
                  [
                    \\"arn:aws:route53:::hostedzone/\\",
                    {
                      \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
                    }
                  ]
                ]
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"PolicyName\\": \\"WebsiteVerifyDomainFunctionServiceRoleDefaultPolicy90A097B8\\",
        \\"Roles\\": [
          {
            \\"Ref\\": \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\"
          }
        ]
      }
    },
    \\"WebsiteVerifyDomainFunction9978DE02\\": {
      \\"Type\\": \\"AWS::Lambda::Function\\",
      \\"Properties\\": {
        \\"Code\\": {
          \\"S3Bucket\\": {
            \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\"
          },
          \\"S3Key\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                },
                {
                  \\"Fn::Select\\": [
                    1,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        \\"Handler\\": \\"index.verifyDomain\\",
        \\"Role\\": {
          \\"Fn::GetAtt\\": [
            \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\",
            \\"Arn\\"
          ]
        },
        \\"Runtime\\": \\"nodejs8.10\\",
        \\"Timeout\\": 60
      },
      \\"DependsOn\\": [
        \\"WebsiteVerifyDomainFunctionServiceRoleDefaultPolicy90A097B8\\",
        \\"WebsiteVerifyDomainFunctionServiceRole77BB87BB\\"
      ]
    },
    \\"WebsiteVerifyDomainB2C2439A\\": {
      \\"Type\\": \\"AWS::CloudFormation::CustomResource\\",
      \\"Properties\\": {
        \\"ServiceToken\\": {
          \\"Fn::GetAtt\\": [
            \\"WebsiteVerifyDomainFunction9978DE02\\",
            \\"Arn\\"
          ]
        },
        \\"HostedZoneId\\": {
          \\"Ref\\": \\"WebsiteHostedZoneD4FAEADF\\"
        },
        \\"DomainName\\": \\"mydomain.com\\",
        \\"CertificateArn\\": \\"arn:aws:acm:REGION:ACCOUNT:certificate/12345-678-9012345\\"
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    }
  },
  \\"Parameters\\": {
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 bucket for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 key for asset version \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168ArtifactHashDB14117A\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"Artifact hash for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    }
  },
  \\"Outputs\\": {
    \\"CloudFrontDistributionId\\": {
      \\"Value\\": {
        \\"Ref\\": \\"WebsiteDistributionCFDistributionD4D90A52\\"
      }
    }
  }
}"
`;
